src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/notifications/NotificationSender.scala

```diff
@@ -24,41 +24,73 @@ import za.co.absa.hyperdrive.trigger.models.{DagInstance, JobInstance, Notificat
                              generalConfig: GeneralConfig) extends NotificationSender {
+  private case class Message(recipients: Seq[String], subject: String, text: String, attempts: Int)
+
   private val logger = LoggerFactory.getLogger(this.getClass)
   private val sender = notificationConfig.senderAddress
   private val notificationEnabled = notificationConfig.enabled
   private val environment = generalConfig.environment
   private val yarnBaseUrl = sparkConfig.hadoopResourceManagerUrlBase
   private val dateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME
+  private val messageQueue = new ConcurrentLinkedQueue[Message]
 
-  def sendNotifications(dagInstance: DagInstance, jobInstances: Seq[JobInstance])(implicit ec: ExecutionContext): Future[Unit] = {
+  def createNotifications(dagInstance: DagInstance, jobInstances: Seq[JobInstance])(implicit ec: ExecutionContext): Future[Unit] = {
     if (notificationEnabled) {
       notificationRuleService.getMatchingNotificationRules(dagInstance.workflowId, dagInstance.status).map {
-        case Some((rules, workflow)) => rules.foreach(rule => createMessageAndSend(rule, workflow, dagInstance, jobInstances))
-        case None => {
-                logger.debug(s"No rules matching workflow ID ${dagInstance.workflowId} with status ${dagInstance.status} found")
-                Future.successful()
-        }
+        case Some((rules, workflow)) => rules
+          .map(rule => createMessage(rule, workflow, dagInstance, jobInstances))
+          .foreach { message =>
+            logger.debug(s"Adding to queue message ${message.subject} from ${sender} to ${message.recipients}")
+            messageQueue.add(message)
+          }
+        case None =>
+          logger.debug(s"No rules matching workflow ID ${dagInstance.workflowId} with status ${dagInstance.status} found")
+          Future.successful()
       }
     } else {
       logger.debug(s"Attempting to send notifications for ${dagInstance}, but it is disabled")
       Future{}
     }
   }
 
-  private def createMessageAndSend(notificationRule: NotificationRule, workflow: Workflow, dagInstance: DagInstance,
-                                   jobInstances: Seq[JobInstance]): Unit = {
+  def sendNotifications(): Unit = {
+    import scala.collection.JavaConverters._
+    val messages = messageQueue.iterator().asScala.toList
+    messageQueue.removeAll(messages.asJava)
+
+    messages.foreach { message =>
+      Thread.sleep(notificationConfig.delay.toMillis)
+      try {
+        logger.debug(s"Sending message ${message.subject} from ${sender} to ${message.recipients}")
+        emailService.sendMessageToBccRecipients(sender, message.recipients, message.subject, message.text)
+      } catch {
+        case e: MailException =>
+          if (message.attempts >= notificationConfig.maxRetries) {
+            logger.error(s"Failed to send message ${message.subject} from ${sender} to ${message.recipients}", e)
+            throw e
```

Comment:
Do nothing instead of throw e ? 
This will fail it for all other messages.

Reply:
Yeah makes sense